### ISSUE TITLE
Android Crash in Fingerprint: "IV has already been used"

### ISSUE THREAD

**Issue by petrfuka** [wultra/powerauth-mobile-sdk#200]
Production crash in Fingerprint dialog.
Crash on old PowerAuth 0.16.5 but issue probably still exists.

Reported stacktrace:
```
Fatal Exception: java.lang.IllegalStateException: IV has already been used. Reusing IV in encryption mode violates security best practices.
       at android.security.keystore.AndroidKeyStoreUnauthenticatedAESCipherSpi.addAlgorithmSpecificParametersToBegin(AndroidKeyStoreUnauthenticatedAESCipherSpi.java:244)
       at android.security.keystore.AndroidKeyStoreCipherSpiBase.ensureKeystoreOperationInitialized(AndroidKeyStoreCipherSpiBase.java:237)
       at android.security.keystore.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:495)
       at javax.crypto.Cipher.doFinal(Cipher.java:1736)
       at io.getlime.security.powerauth.keychain.fingerprint.FingerprintAuthenticationHandler.encryptedKey(FingerprintAuthenticationHandler.java:236)
       at io.getlime.security.powerauth.keychain.fingerprint.FingerprintAuthenticationDialogFragment$5.run(FingerprintAuthenticationDialogFragment.java:331)
       at android.os.Handler.handleCallback(Handler.java:789)
       at android.os.Handler.dispatchMessage(Handler.java:98)
       at android.os.Looper.loop(Looper.java:164)
       at android.app.ActivityThread.main(ActivityThread.java:6944)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
```

**Comment by hvge** [comment]
Do you have more information about the device? This might be a device specific issue.

**Comment by hvge** [comment]
Looks like we somehow call `FingerprintAuthenticationHandler.encryptedKey()` for twice. That leads to two subsequent calls to `javax.crypto.Cipher.doFinal()` and the underlying implementation throws that exception. Normally, `doFinal()` should leave `Cipher` object to its after-init state, but this is Android ✌️[Source of the exception](https://android.googlesource.com/platform/frameworks/base/+/master/keystore/java/android/security/keystore/AndroidKeyStoreUnauthenticatedAESCipherSpi.java#240)

I think the best solution is to re-initialize the Cipher from scratch for each encryption, or ensure that UI (dialog fragment) doesn't call that `encryptedKey()` for twice.

**Comment by hvge** [comment]
I just found out that this fix now causes that the unlock key for the biometric factor is wrongly calculated. The derived key is always the same, with the following implications:

- If activation was created in older version of SDK, then the biometric authentication will not work properly
- If activation was created in `0.20.1` (or `1.1.0`), then the weak key is protecting that factor.

We recommend you to do not use version `0.20.1`  (or `1.1.0`) on Androids, until we prepare hotfix.

**Comment by hvge** [comment]
OK, I just released hotfixed version `0.20.2`. For `1.1.0` release (which is currently used only by Wultra), I'll prepare a proper fix for this issue, which will be backported to `0.20.x` branch.
